KEYCLOAK-12278 Default first broker login flow is broken after migration

mposolda opened the pull request:
<!---
Please read https://github.com/keycloak/keycloak/blob/master/CONTRIBUTING.md and follow these guidelines when contributing to Keycloak
-->

mposolda committed:
KEYCLOAK-12278 Default first broker login flow is broken after migration

mposolda commented:
Documentation PR: https://github.com/keycloak/keycloak-documentation/pull/793

mposolda commented:
@stianst @hmlnarik @pedroigor Anyone of you has a chance to review this? Thanks

stianst commented:
If we're doing a 8.0.2 release then this would need to be reworked since it contains migration for 9.

thielef commented on server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java:
```diff
@@ -83,4 +89,71 @@ protected void addAccountConsoleClient(RealmModel realm) {
+
+
+    protected void migrateAuthenticationFlowsWithAlternativeRequirements(RealmModel realm) {
+        for (AuthenticationFlowModel flow : realm.getAuthenticationFlows()) {
```
Is this also returing non-root flows (e.g. OTP sub flow of the browser flow)?

mposolda replied:
yes, it is returning also non-root flows.

thielef commented on server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java:
```diff
@@ -83,4 +89,71 @@ protected void addAccountConsoleClient(RealmModel realm) {
+            boolean requiredFound = false;
+            LinkedList<AuthenticationExecutionModel> alternativesAfterRequired = new LinkedList<>();
+            for (AuthenticationExecutionModel execution : realm.getAuthenticationExecutions(flow.getId())) {
+                if (AuthenticationExecutionModel.Requirement.REQUIRED.equals(execution.getRequirement())) {
```
For a flow like...
- 1) ExecutionA, conditional
- 2) ExecutionB, alternative
- 3) ExecutionC, alternative
...there would be no migration. But effectively, this flow is now ignoring step 2 and 3 as no migration is applied (due to org.keycloak.authentication.DefaultAuthenticationFlow.processFlow():
`            else if (execution.isRequired() || execution.isConditional()) {
                requiredList.add(execution);
            }
`

mposolda replied:
You're right. Will improve that.

thielef commented on server-spi-private/src/main/java/org/keycloak/migration/migrators/MigrateTo9_0_0.java:
```diff
@@ -83,4 +89,71 @@ protected void addAccountConsoleClient(RealmModel realm) {
+                }
+
+                // ALTERNATIVE executions found after REQUIRED. We may migrate them
+                if (AuthenticationExecutionModel.Requirement.ALTERNATIVE.equals(execution.getRequirement()) && requiredFound) {
```
For a flow like...
- ExecutionA, alternative
- ExecutionB, alternative
- ExecutionC, required
...there would be no migration either. But effectively, this flow is now ignoring step 1 and 2 as no migration is applied (see mail thread on dev mailing list -- our custom SPI use case).

mposolda replied:
You're right. Will improve that.

mposolda commented:
@stianst @thielef I've addressed some concerns and changed this PR to use MigrateTo8_0_2 . The commit should be backportable to 8.0.2 branch without issues. Do you have a chance to re-review?

Also changed the documentation PR and update the version there to 8.0.2 : https://github.com/keycloak/keycloak-documentation/pull/793

thielef commented:
👍 . Looks fine now (especially the fact that we get it with 8.0.2 now :-) )

thielef commented:
👍 . Looks fine now (especially the fact that we get it with 8.0.2 now :-) )

mposolda commented:
@thielef Thanks!

stianst reviewed (commented):
Looks okay to me, but have only done a brief review.

stianst commented:
@mposolda I can switch to approve once someone has done a deeper review of this

mposolda commented:
@stianst Thanks, @iankko will take a look. Also the @thielef did good job with reviewing this. Just updated this to fix the rebase conflict

iankko reviewed (approved):
LGTM (see below for more testing details / comparison of the migration script behaviour across various versions)

Changes in the behaviour of the migration script (from local testing, when using MySQL as the DB backend, and GitHub as the user identity broker) are as follows:

1. With Keycloak v7.0.1, define GitHub identity broker & import users. 
2. Stop the v7.0.1 server version, copy content of _standalone/_ to v8.0.0 server version, and run the _migrate-standalone.cli_ script. Start the v8.0.0 server (with properties defining the connection to the MySQL backend). Server starts successfully, but once the (previously) imported users are deleted, and tried to be created again (via broker login), the following screen appears:

![we_are_sorry](https://user-images.githubusercontent.com/8414918/71278044-cbb4ba00-2356-11ea-89a3-17bbc43376ed.png)

and message like the following appears in the Keycloak v8.0.0 server log:
```
17:55:15,254 WARN  [org.keycloak.connections.httpclient.DefaultHttpClientFactory] (default task-2) TruststoreProvider is disabled
17:58:03,957 WARN  [org.keycloak.services] (default task-1) KC-SERVICES0013: Failed authentication: org.keycloak.authentication.AuthenticationFlowException
...
17:58:03,960 WARN  [org.keycloak.events] (default task-1) type=IDENTITY_PROVIDER_FIRST_LOGIN_ERROR, realmId=master, clientId=account, userId=null, ipAddress=127.0.0.1, error=invalid_user_credentials, identity_provider=github, auth_method=openid-connect, redirect_uri=http://localhost:8080/auth/realms/master/account/login-redirect, identity_provider_identity=iankko, code_id=d9c90724-2a7a-4436-a03a-86e3c6179717, authSessionParentId=d9c90724-2a7a-4436-a03a-86e3c6179717, authSessionTabId=_hdhRgqgWGI
```

3. Stop the v8.0.0 of the Keycloak server. Start the v7.0.1 of the Keycloak server again & import users from the identity broker again (by first log in them). Stop the v7.0.1 version.
4. Copy the content of the v7.0.1 _standalone/_ directory to v8.0.2 / v9.0.0 (the Keycloak server version with change from this PR included), and perform the `migrate-standalone.cli` script again. Start the v9.0.0 version of the server (with corresponding MySQL backend system properties). Server starts properly. Previously imported users are present. Once deleted, they are imported successfully again (upon first login via the social broker). No error message compared to the case, when migrating from v7.0.1 to v8.0.0 is present (IOW everything works fine).
5. **Moreover (bonus points)**, seems it's even possible to recover from the _"broken"_ v8.0.0 migrated configuration into the _"working"_ v9.0.0 configuration. Stop the v9.0.0 version of the server. Copy the _standalone/_ content from v8.0.0 to v9.0.0, and run the `./jboss-cli --file=migrate-standalone.cli` snippet again. Start the v9.0.0 (or v8.0.2 if you want to call it that way) version of the Keycloak server. Server starts properly, identity broker user can login properly. Once deleted (via admin), and logged via the broker for the 1-th time again, the users are present in the database (and in the UI via "_View all users_") again.

mposolda commented:
@iankko Thanks for the review and for the additional detailed testing!
